[6.2] Styles and Script element in rendering Debug Assets

sinahaghparast opened the pull request:
Pull Request for Issue # .

### Summary of Changes
Code be better...


### Testing Instructions
[https://validator.w3.org/](https://validator.w3.org/)


### Actual result BEFORE applying this Pull Request
> Validator message: 
> Info: The type attribute is unnecessary for JavaScript resources.
> Info: Trailing slash on void elements has no effect and interacts badly with unquoted attribute values.


### Expected result AFTER applying this Pull Request
> Validator message: 
> Document checking completed. No errors or warnings to show.


### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

sinahaghparast committed:
Update StylesRenderer.php

sinahaghparast committed:
Update JavascriptRenderer.php

sinahaghparast committed:
Update hits.php

sinahaghparast committed:
Update default_items.php

sinahaghparast committed:
Update default.php

brianteeman commented:
Before jumping in and "fixing" the code you really need to spend a little more time both using Joomla dn looking through the code history.

For example you will find that the "itemprop" you are adding here has only just been removed because we switched to use json rich snippets see #41151 so adding it here is completely wrong.

Secondly a pull request should address a single issue. Here you are addressing two unrelated issues

sinahaghparast committed:
Update default.php

sinahaghparast committed:
Update default_items.php

sinahaghparast committed:
Update hits.php

sinahaghparast commented:
I removed itemprop...

brianteeman commented:
> I removed itemprop...

Now you need to make sure the pull request title and description are still correct and if not then you will need to edit those. The title is particularly important as that is what people will seein the changelog

brianteeman commented:
the javascript renderer for the debug plugin is not in the head. It is just before the closing body tag. You probably should also fix the css for that plugin as well. 

![image](https://github.com/joomla/joomla-cms/assets/1296369/c85e74b4-2fe1-4b1d-a6ca-d98874722ef2)

dgrammatiko commented:
The code for the debug css is here: https://github.com/joomla/joomla-cms/blob/45adb61f8306b945f43b4326305fafe87c154aa2/plugins/system/debug/src/JavascriptRenderer.php#L63
For the JS is here: https://github.com/joomla/joomla-cms/blob/45adb61f8306b945f43b4326305fafe87c154aa2/plugins/system/debug/src/JavascriptRenderer.php#L71

Also just reviewing the code the nonce attribute needs to be applied for each stylesheet/script, right now debug is probably broken with CSP enabled

sinahaghparast committed:
Update JavascriptRenderer.php

sinahaghparast commented:
I apologize for the delay. It was related to my personal life.

According to this [document](https://content-security-policy.com/nonce/), external files can also have the nonce attribute. what is your opinion? be added?

sinahaghparast commented:
Also, I don't agree with adding the nonce attribute to the inline style. Because according to this [document](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src), the code inside the elements are also blocked.
> Inline style attributes are also blocked:
> `<div style="display:none">Foo</div>`

sinahaghparast committed:
Update JavascriptRenderer.php

dgrammatiko commented:
> Also, I don't agree with adding the nonce attribute to the inline style. Because according to this [document](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src), the code inside the elements are also blocked.

You misunderstood what the inline style it refers to. It's not the `<style>` tag but rather the attribute `style="display: none"`.

<img width="441" alt="Screenshot 2024-04-16 at 6 48 11 PM" src="https://github.com/joomla/joomla-cms/assets/3889375/73745567-3ee8-4266-b653-99f4f96700e1">


FWIW Joomla is already adding the nonce for the head inline css/js tags.

dgrammatiko commented on plugins/system/debug/src/JavascriptRenderer.php:
```diff
@@ -59,22 +59,21 @@ public function renderHead()
-
-        if ($doc->cspNonce) {
-            $nonce = ' nonce="' . $doc->cspNonce . '"';
+            $html .= sprintf('<script src="%s" defer></script>' . "\n", $file);
```
```suggestion
            $html .= sprintf('<script src="%s" defer%s></script>' . "\n", $file, $nonce);
```

sinahaghparast commented:
If I have misunderstood, then you guide me so that I can be informed. Just stating that I got it wrong won't help me. It may be repeated until the offender realizes his mistake. Please introduce new document or interpret that document.

sinahaghparast committed:
Update JavascriptRenderer.php

sinahaghparast commented:
I read a page from the MDN website. And I understood such a concept. Anyway, I trust your words. Are the current changes acceptable?

sinahaghparast committed:
Update StylesRenderer.php

Quy commented on libraries/src/Document/Renderer/Html/StylesRenderer.php:
```diff
@@ -194,10 +194,10 @@ private function renderElement($item): string
 
         if ($relation === 'lazy-stylesheet') {
-            $buffer .= '<noscript><link href="' . htmlspecialchars($src) . '" rel="stylesheet" /></noscript>';
+            $buffer .= '<noscript><link href="' . htmlspecialchars($src) . '" rel="stylesheet" ></noscript>';
```
```suggestion
            $buffer .= '<noscript><link href="' . htmlspecialchars($src) . '" rel="stylesheet"></noscript>';
```

sinahaghparast committed:
Update StylesRenderer.php

brianteeman commented on plugins/system/debug/src/JavascriptRenderer.php:
```diff
@@ -59,22 +59,21 @@ public function renderHead()
 
         foreach ($inlineCss as $content) {
-            $html .= sprintf('<style>%s</style>' . "\n", $content);
+            $html .= sprintf('<style%s>%s</style>' . "\n", $nonce, $content);
```
???

sinahaghparast replied:
In two webpage wrote about style and nonce...

[https://content-security-policy.com/nonce/](https://content-security-policy.com/nonce/)
![Screenshot 2024-05-15 at 00-04-37 CSP Nonce ⟶ Script   Style Attribute](https://github.com/joomla/joomla-cms/assets/33615396/9e92c886-ac10-444b-8c95-22a3d4c607fb)

[https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src#unsafe_inline_styles](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src#unsafe_inline_styles)

Please read them and make the final decision with the experienced people of the community.
My personal opinion is not to use nonce for the style tag.

sinahaghparast replied:
Before I wrote,
[https://github.com/joomla/joomla-cms/pull/43239#issuecomment-2059406886](https://github.com/joomla/joomla-cms/pull/43239#issuecomment-2059406886)

And @dgrammatiko [here](https://github.com/joomla/joomla-cms/pull/43239#issuecomment-2059415686) proposed "...adding the nonce for the head inline css/js tags."

dgrammatiko replied:
The line is correct

brianteeman replied:
Shouldnt it be
```suggestion
            $html .= sprintf('<style%1$s>%2$s</style>' . "\n", $nonce, $content);
```

sinahaghparast replied:
It does not matter which method is used here.
For example, If we have be two values and they are to be placed in three positions, your suggested method should be implemented.

`$formattedString = sprintf("full name: %1\$s %2\$s first name: %1\$s last name: %2\$s", $firstName, $lastName);`

sinahaghparast committed:
Merge branch '5.2-dev' into 5.2-StylesRenderer.php-JavascriptRenderer.php

angieradtke commented:
trailing slash and needless attribute in  debug bar  removed :-)

angieradtke commented:
I have tested this item :white_check_mark: successfully on b609e14ef386f863502465a5317386eeea1a79d7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43239">issues.joomla.org/tracker/joomla-cms/43239</a>.</sub>

brianteeman commented on plugins/system/debug/src/JavascriptRenderer.php:
```diff
@@ -59,22 +59,21 @@ public function renderHead()
-
-        if ($doc->cspNonce) {
-            $nonce = ' nonce="' . $doc->cspNonce . '"';
+            $html .= sprintf('<script src="%s" defer%s></script>' . "\n", $file, $nonce);
```
what is the purpose of the %s after defer?
it doesnt appear to do anything

richard67 replied:
It will add ` nonce="xyz"` after the `defer` if `$nonce` is not an empty string, i.e. if we have a CSP nonce. See lines 62 to 65 how `$nonce` is set.

richard67 replied:
Was added with a code suggestion by Dimitris here: https://github.com/joomla/joomla-cms/pull/43239#discussion_r1567609062 .

HLeithner commented:
This pull request has been automatically rebased to 5.3-dev.

richard67 committed:
Merge branch '5.3-dev' into 5.2-StylesRenderer.php-JavascriptRenderer.php

richard67 commented on plugins/system/debug/src/JavascriptRenderer.php:
```diff
@@ -59,22 +59,21 @@ public function renderHead()
-
-        if ($doc->cspNonce) {
-            $nonce = ' nonce="' . $doc->cspNonce . '"';
+            $html .= \sprintf('<script src="%s" defer%s></script>' . "\n", $file, $nonce);
```
@brianteeman 's previous question in his comment here https://github.com/joomla/joomla-cms/pull/43239#discussion_r1730026560 is still valid: What is the purpose of the %s after defer? @sinahaghparast Please check.

richard67 replied:
Hmm, I just see it will add ` nonce="xyz"` after the `defer` if `$nonce` is not an empty string, i.e. if we have a CSP nonce.

sinahaghparast committed:
Update JavascriptRenderer.php

My third child was born recently. That's why I sit at the computer a little less. I hope the final changes are satisfactory.

sinahaghparast committed:
Update JavascriptRenderer.php

sinahaghparast commented:
@brianteeman 
title and code edited.

sinahaghparast committed:
Update JavascriptRenderer.php

brianteeman commented:
> @brianteeman title and code edited.

the title still says this is for the **head**

sinahaghparast commented:
> > @brianteeman title and code edited.
> 
> the title still says this is for the **head**

@brianteeman 
**renderHead** is function name...
What do you recommend?

dgrammatiko commented:
@sinahaghparast the debugbar injects the assets at the end of the body (although the class says head) so maybe instead of `head` should say `debugbar`?

richard67 committed:
Merge branch '5.3-dev' into 5.2-StylesRenderer.php-JavascriptRenderer.php

sinahaghparast commented:
> > > @brianteeman title and code edited.
> > 
> > 
> > the title still says this is for the **head**
> 
> @brianteeman **renderHead** is function's name... What do you recommend?

I suggest to name the function "renderDebugAssets". 
Is okay?

richard67 commented:
> I suggest to name the function "renderDebugAssets". Is okay?

@sinahaghparast I would not change the method name for b/c (backwards compatibility) reasons. But the title of your pull request is better now.

HLeithner commented:
This pull request has been automatically rebased to 6.0-dev.

brianteeman commented:
are you able to resolve the conflicts and then we can get this merged please

richard67 committed:
Merge branch '6.0-dev' into 5.2-StylesRenderer.php-JavascriptRenderer.php

richard67 commented:
I've allowed myself to fix the conflict.

brianteeman commented:
I have tested this item :white_check_mark: successfully on 08d60b5f3c9e44ba41a72b827012af93eadfdfc7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43239">issues.joomla.org/tracker/joomla-cms/43239</a>.</sub>

sinahaghparast committed:
Update JavascriptRenderer.php

HLeithner commented:
This pull request has been automatically rebased to 6.1-dev.
